abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,919 @@
+
+All functions implemented in the Wasm module, other than the integration and memory management
+functions, include context identifier (`context_id`) as the first parameter, which should be used to
+distinguish between different contexts.
+
+
+## Integration
+
+### `_start`
+
+* params:
+  - none
+* returns:
+  - none
+
+Start function which is called when the module is loaded and initialized. This can be used by SDKs
+to setup and/or initialize state, but no proxy_ functions can be used at that point yet.
+
+
+### `proxy_abi_version_X_Y_Z`
+
+* params:
+  - none
+* returns:
+  - none
+
+Exports ABI version (vX.Y.Z) this module was compiled against.
+
+Considered alternatives: Ideally, this would be exported as a global variable, but existing
+toolchains have trouble with that (some don’t support global variables at all, some incorrectly
+export language-specific and not Wasm-generic variables).
+
+
+## Memory management
+
+### `proxy_on_memory_allocate`
+
+* params:
+  - `i32 (size_t) memory_size`
+* returns:
+  - `i32 (void*) allocated_ptr`
+
+Allocates memory using in-VM memory allocator and returns it to the host.
+
+
+## Module lifecycle
+
+### `proxy_on_context_create`
+
+* params:
+  - `i32 (uint32_t) context_id`
+  - `i32 (uint32_t) parent_context_id`
+* returns:
+  - none
+
+Called when the host environment creates a new root context (if `parent_context_id` is `0`) or a new
+per-stream context.
+
+
+### `proxy_on_done`
+
+* params:
+  - `i32 (uint32_t) context_id`
+* returns:
+  - `i32 (bool) is_done`
+
+Called when the host environment is done processing the context (`context_id`). Return value
+indicates when the Wasm VM is done with the processing as well.
+
+
+### `proxy_on_log`
+
+* params:
+  - `i32 (uint32_t) context_id`
+* returns:
+  - none
+
+Called in the logging phase.
+
+
+### `proxy_on_delete`
+
+* params:
+  - `i32 (uint32_t) context_id`
+* returns:
+  - none
+
+Called when the host environment removes the context (`context_id`). This is used to signal that VM
+should stop tracking that `context_id` and remove all associated state.
+
+
+## Configuration
+
+### `proxy_on_vm_start`
+
+* params:
+  - `i32 (uint32_t) root_context_id`
+  - `i32 (size_t) vm_configuration_size`
+* returns:
+  - `i32 (bool) success`
+
+Called when the host environment starts the WebAssembly Virtual Machine. Its configuration
+(`vm_configuration_size`) might be retrieved using `proxy_get_buffer`.
+
+
+### `proxy_on_configure`
+
+* params:
+  - `i32 (uint32_t) root_context_id`
+  - `i32 (size_t) plugin_configuration_size`
+* returns:
+  - `i32 (bool) success`
+
+Called when the host environment starts the plugin. Its configuration (`plugin_configuration_size`)
+might be retrieved using `proxy_get_buffer`.
+
+
+## Timers
+
+### `proxy_on_tick`
+
+* params:
+  - `i32 (uint32_t) root_context_id`
+* returns:
+  - none
+
+Timer called every tick period. Tick period can be configured using
+`proxy_set_tick_period_milliseconds`.
```

Comment:
No default, there is no ticking unless configured.

Reply:
We could add a token_ptr and if the target token was zero we could use that to indicate that the system should allocate a new timer, but if it was non-zero then it would be the timer to set a new period for.  WDYT?  This would allow multiple timers.

Reply:
+1 for some mechanism to have multiple timers. Also it's not clear to me how to disable a timer so we should document that. I guess set a tick period to 0 for a particular timer token?

Reply:
yes, that is correct.  we also need to add a uint32_t token to the on_tick call into the VM.

Reply:
+1